diff -Nur -x '*.orig' -x '*.rej' e-smith-portforwarding-1.2.0/root/etc/e-smith/templates/etc/rc.d/init.d/masq/91adjustPortForward mezzanine_patched_e-smith-portforwarding-1.2.0/root/etc/e-smith/templates/etc/rc.d/init.d/masq/91adjustPortForward
--- e-smith-portforwarding-1.2.0/root/etc/e-smith/templates/etc/rc.d/init.d/masq/91adjustPortForward	2007-10-16 11:46:31.000000000 -0400
+++ mezzanine_patched_e-smith-portforwarding-1.2.0/root/etc/e-smith/templates/etc/rc.d/init.d/masq/91adjustPortForward	2007-10-16 11:45:51.000000000 -0400
@@ -21,7 +21,7 @@
             $port =~ s/-/:/;
 
 	    # Map canonical localhost back to our current external IP	
-            $ip = $ExternalInterface{IPAddress} if ($ip eq 'localhost');
+            $ip = '$OUTERNET' if ($ip eq 'localhost');
 
             $OUT .= "    /sbin/iptables --table nat --append $pf_chain " .
                 "--protocol $protocol \\\n".
@@ -39,7 +39,7 @@
             # then we must allow it on the INPUT chain instead of the FORWARD
             # chain.
             $OUT .= "    adjust_${protocol}_in $port ACCEPT " .
-			(($ip eq $ExternalInterface{IPAddress}) ?
+			(($ip eq '$OUTERNET') ?
 			    "Inbound${uproto}_\$\$\n" :
 			    "Forwarded${uproto}_\$\$ $ip/32\n");
         }
