diff -Nur -x '*.orig' -x '*.rej' e-smith-portforwarding-1.2.0/root/etc/e-smith/locale/en-us/etc/e-smith/web/functions/portforwarding mezzanine_patched_e-smith-portforwarding-1.2.0/root/etc/e-smith/locale/en-us/etc/e-smith/web/functions/portforwarding
--- e-smith-portforwarding-1.2.0/root/etc/e-smith/locale/en-us/etc/e-smith/web/functions/portforwarding	2008-02-13 23:55:25.000000000 +1100
+++ mezzanine_patched_e-smith-portforwarding-1.2.0/root/etc/e-smith/locale/en-us/etc/e-smith/web/functions/portforwarding	2008-02-13 23:54:48.000000000 +1100
@@ -86,44 +86,20 @@
         <trans>Destination Host IP Address</trans>
     </entry>
     <entry>
-        <base>Configuration</base>
-        <trans>Configuration</trans>
-    </entry>
-    <entry>
         <base>Port forwarding</base>
         <trans>Port forwarding</trans>
     </entry>
     <entry>
-        <base>Security</base>
-        <trans>Security</trans>
-    </entry>
-    <entry>
-        <base>SAVE</base>
-        <trans>Save</trans>
-    </entry>
-    <entry>
         <base>APPLY</base>
         <trans>Add</trans>
     </entry>
     <entry>
-        <base>NEXT</base>
-        <trans>Next</trans>
-    </entry>
-    <entry>
-        <base>CANCEL</base>
-        <trans>Cancel</trans>
-    </entry>
-    <entry>
         <base>SUCCESS</base>
         <trans>Your change to the port forwarding rules has been
         successfully saved.
         </trans>
     </entry>
     <entry>
-        <base>REMOVE</base>
-        <trans>Remove</trans>
-    </entry>
-    <entry>
         <base>ERR_NO_MASQ_RECORD</base>
         <trans>Cannot retrieve masq record from the configuration
         database.</trans>
@@ -157,10 +133,6 @@
         twice.</trans>
     </entry>
     <entry>
-        <base>ACTION</base>
-        <trans>Action</trans>
-    </entry>
-    <entry>
         <base>ERR_PORT_COLLISION</base>
         <trans>
             ERROR: This port or port range conflicts with an existing
